fix(vault): secret is temporarily empty after changed vault config

cshuaimin opened the pull request:
<!--
NOTE: Please read the CONTRIBUTING.md guidelines before submitting your patch,
and ensure you followed them all:
https://github.com/Kong/kong/blob/master/CONTRIBUTING.md#contributing

Refer to the Kong Gateway Community Pledge to understand how we work
with the open source community:
https://github.com/Kong/kong/blob/master/COMMUNITY_PLEDGE.md
-->

### Summary

<!--- Why is this change required? What problem does it solve? -->
When vault config is changed, in the worker event we first flush the LRU cache, then start to update the secrets from vault provider. There’s a period of time in between that the cache is empty.
The `kong.vault.update()` function only lookups cache and will update the secret to an empty string when cache is empty. This can cause plugins to throw nil errors.
This commit changed `kong.vault.update()` function to not touch it if not found in the cache.
### Checklist

- [x] The Pull Request has tests
- [x] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/blob/main/README.md)
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Issue reference

<!--- If it fixes an open issue, please link to the issue here. -->
FTI-5936

cshuaimin committed:
fix(vault): secret is temporarily empty after changed vault config

When vault config is changed, in the worker event we first flush the LRU
cache, then start to update the secrets from vault provider. There’s a
period of time in between that the cache is empty.
The `kong.vault.update()` function only lookups cache and will update
the secret to an empty string when cache is empty. This can cause plugins
to throw nil errors. This commit changed `kong.vault.update()` function
to not touch it if not found in the cache.

cshuaimin committed:
fix(vault): secret is temporarily empty after changed vault config

When vault config is changed, in the worker event we first flush the LRU
cache, then start to update the secrets from vault provider. There’s a
period of time in between that the cache is empty.
The `kong.vault.update()` function only lookups cache and will update
the secret to an empty string when cache is empty. This can cause plugins
to throw nil errors. This commit changed the worker event callback to
not flush the LRU cache. The cache will be updated when the secrets are
fetched from vault prodiver.

cshuaimin commented:
Pushed a new commit to fix the issue in another way: instead of not touching input table in the pdk update function, in this new commit I completely removed the LRU:flush_all() line in the worker event. The LRU capacity is fixed and the keys has ttls, so there likely won’t be memory leaks.

bungle commented on kong/pdk/vault.lua:
```diff
@@ -1429,8 +1429,6 @@ local function new(self)
     end
 
-    LRU:flush_all()
-
```
The `LuaDoc` has:

```
  ---
  -- Flushes LRU caches and forcibly rotates the secrets.
  --
  -- This is only ever executed on traditional nodes.
  ```

I guess this may need to be changed. Also should it be executed with incremental too?

cshuaimin commented:
Hi @bungle sorry for the late reply, I just took another look at this. I think current approach may not be ideal for these reasons:

1. It's not enough to completely fix the 'could not find cached value' issue, because there's another lru flush in kong.vault.flush() function. In my test, I changed the vault config in Konnect UI and can still see several such errors in a short time even after this change.  And I think we can not remove the flush in this place right? Because the function is called flush...
2. the configuration stickyness issue may come back (not very sure)

Do you think the first commit is okay? The only downside is the behavior of kong.vault.flush() changed slightly. But the function is marked as experimental in the docs, and seems that the usage of the function in kong will not be affected. If it's acceptable, I will revert the second commit and change the failed test "update function sets values to empty string on failure".

cshuaimin committed:
Back out "fix(vault): secret is temporarily empty after changed vault config"

This backs out commit 4f7cf0970738f5714569aa4450b68b05d7731894.

bungle commented:
@cshuaimin The first commit is ok with me! Thanks!

cshuaimin committed:
tests(vault): kong.vault.update() now doesn't change values on failure

cshuaimin committed:
fix the update logic to change values if no error

The test "resurrects plugin config references when secret is deleted" failed.
It is asserting that when a secret resurrect ttl is passed, it should be
deleted. However with previous update() function logic the value is
only set if it's not nil. This commit changes the logic to set the value
based on if there's an error or not.
